20230729-db-cloud-auth.md

```diff
@@ -0,0 +1,172 @@
+This requires a background process to refresh the token within application.
+
+#### GCP
+
+GCP Cloud SQL Go Connector authentication is done via [a special dialer](https://cloud.google.com/sql/docs/postgres/samples/cloud-sql-postgres-databasesql-connect-connector).
+This method is simpler than AWS RDS IAM, since it does not require a background process to refresh the token. I think [go-cloud](https://github.com/google/go-cloud)
+provides this functionality out of the box, so that could be used. 
+
+### Database connectors
+
+Below are some examples of how the connectors could be implemented with AWS RDS IAM for both postgres and mysql.
+
+#### Postgres
+
+Since openFGA uses pgx, we can use the [BeforeConnect](https://github.com/jackc/pgx/blob/d626dfe94e250e911b77ac929e2be06f81042bd0/pgxpool/pool.go#L109)
+hook to provide AWS RDS IAM tokens.
+
+```go
+func openRDS(uri *url.URL, cfg *sqlcommon.Config) (*sql.DB, error) {
+	ctx := context.Background()
+	iam, err := newIAM(ctx, cfg.Username, uri.Hostname(), uri.Port())
+	if err != nil {
+		return nil, err
+	}
+	conf, err := pgx.ParseConfig(uri.String())
+	if err != nil {
+		return nil, err
+	}
+	connector := pgxstd.GetConnector(*conf, pgxstd.OptionBeforeConnect(iam.BeforeConnect))
+	db := sql.OpenDB(connector)
+	return db, nil
+}
+```
+
+#### Mysql
+
+For mysql we need to wrap the sql connector to modify the connection string. This is done by implementing the [Driver interface](https://golang.org/pkg/database/sql/driver/#Driver).
+
+```go
+func (i *iam) Connect(ctx context.Context) (driver.Conn, error) {
+	token, isNew, err := i.getRDSAuthToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if !isNew {
+		return i.baseConnector.Connect(ctx)
+	}
+	i.mysqlCfg.Passwd = token
+	i.baseConnector, err = mysql.NewConnector(i.mysqlCfg)
+	if err != nil {
+		return nil, err
+	}
+	return i.baseConnector.Connect(ctx)
+}
+
+func (i *iam) Driver() driver.Driver {
+	return i.baseConnector.Driver()
+}
+
+func open(...) (*sql.DB, error) {
+	i := newIAM(...)
+	return sql.OpenDB(i), nil
+}
+```
+## Extensibility
+
+This implementation would make it relatively simple to add new providers. You would just add a new case to `auth_method` and implement the connector.
+
+# Migration
+[migration]: #migration
+
+There should not be any breaking changes just new configuration options.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+ - Adds dependencies to cloud provider SDKs.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+- Why is this proposal the best?
+- What is the impact of not doing this?
+
+# Prior Art
+[prior-art]: #prior-art
+
+### [go-cloud](https://github.com/google/go-cloud) 
+
+Provides [IAM auth for GCP Cloud SQL](https://github.com/google/go-cloud/blob/master/postgres/gcppostgres/gcppostgres_test.go#L54), but not [AWS RDS](https://github.com/google/go-cloud/issues/3069). 
```

Comment:
go-cloud does support AWS RDS. This issue is outdated and has since been closed.

See https://gocloud.dev/howto/sql/#aws

Reply:
There some support for RDS, but it is not complete (see my other comment), final comment on the issue also implied that they did not intend to add the [IAM auth](https://github.com/google/go-cloud/issues/3069#issuecomment-929512936)